### ISSUE TITLE
Training with the same parameters and seed gets significantly different results

### ISSUE THREAD

**Issue by fedorzh** [apache/incubator-mxnet#9410]
Train with the same parameters and seeds, get different results. Repro:

```

# coding: utf-8

# In[1]:


import mxnet as mx
from mxnet import nd, gluon, autograd, ndarray
import numpy as np
import random

def transform(data, label):
    return [dat.astype(np.float32) for dat in data], [lab.astype(np.float32) for lab in label]


train_cifar_gluon = gluon.data.vision.CIFAR10(train=True, transform=transform)
test_cifar_gluon = gluon.data.vision.CIFAR10(train=False, transform=transform)

def convert_gluon_dataset_to_numpy(data):
    ds = data[:][0][0].shape
    X = np.empty((len(data[:][0]), ds[2], ds[0], ds[1]), dtype=np.float32)
    for i, example in enumerate(data[:][0]):
        X[i, :] = np.rollaxis(example.asnumpy(),2)
    y = np.array(data[:][1])
    return X, y

X, y = convert_gluon_dataset_to_numpy(train_cifar_gluon)
X_test, y_test = convert_gluon_dataset_to_numpy(test_cifar_gluon)


# In[2]:


def predict_scores(net, X_, batch_size, context):
    scores = None
    test_loaded = gluon.data.DataLoader(mx.nd.array(X_), batch_size, shuffle=False)
    for data in test_loaded:
        data = data.as_in_context(context)
        output = net(data).asnumpy()
        if scores is None:
            scores = output
        else:
            scores = np.append(scores, output, axis=0)
    return scores


# In[3]:


gpu_count = 1
_ctx_list = [mx.gpu(i) for i in range(gpu_count)]
_batch_size=64
epochs=1
_seed=42
_optimizer='sgd'
_learning_rate=0.1
_xavier_magnitude=2.
_momentum=0.9
_wd=0.0001
_nclasses=10


# ### Try 1

# In[4]:


random.seed(_seed)
mx.random.seed(_seed)
np.random.seed(_seed)


# In[5]:


net = gluon.model_zoo.vision.get_model('resnet34_v2', pretrained=False, classes=_nclasses, ctx=_ctx_list)

loss = gluon.loss.SoftmaxCrossEntropyLoss()


# In[6]:


net.collect_params().initialize(mx.init.Xavier(magnitude=_xavier_magnitude), ctx=_ctx_list, force_reinit=True)

trainer = gluon.Trainer(net.collect_params(), _optimizer,
                        optimizer_params=dict(learning_rate=_learning_rate, momentum=_momentum,
                                              wd=_wd),
                        kvstore='device' if len(_ctx_list) > 0 else 'local')

train_data = mx.io.NDArrayIter(X, label=y, batch_size=_batch_size)
for e in range(epochs):
    train_data.reset()
    for batch in train_data:
        cur_contexts = _ctx_list
        if batch.data[0].shape[0] < len(_ctx_list):
            cur_contexts = cur_contexts[:batch.data[0].shape[0]]
        data = gluon.utils.split_and_load(batch.data[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        label = gluon.utils.split_and_load(batch.label[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        Ls = []
        with autograd.record():  # Start recording the derivatives
            for x_cur, y_cur in zip(data, label):
                L = loss(net(x_cur), y_cur)
                # store the loss and do backward after we have done forward
                # on all GPUs for better speed on multiple GPUs.
                Ls.append(L)
            for L in Ls:
                L.backward()
        trainer.step(batch.data[0].shape[0])
        
    scores_test = predict_scores(net, X_test, _batch_size, _ctx_list[0])
    predictions_test = np.argmax(scores_test, axis=1)
    accuracy = np.mean(predictions_test == y_test)
    print('[Epoch %d] accuracy=%f' % (e, accuracy))


# ### Try 2

# In[7]:


random.seed(_seed)
mx.random.seed(_seed)
np.random.seed(_seed)


# In[8]:


net = gluon.model_zoo.vision.get_model('resnet34_v2', pretrained=False, classes=_nclasses, ctx=_ctx_list)

loss = gluon.loss.SoftmaxCrossEntropyLoss()


# In[9]:


net.collect_params().initialize(mx.init.Xavier(magnitude=_xavier_magnitude), ctx=_ctx_list, force_reinit=True)

trainer = gluon.Trainer(net.collect_params(), _optimizer,
                        optimizer_params=dict(learning_rate=_learning_rate, momentum=_momentum,
                                              wd=_wd),
                        kvstore='device' if len(_ctx_list) > 0 else 'local')

train_data = mx.io.NDArrayIter(X, label=y, batch_size=_batch_size)
for e in range(epochs):
    train_data.reset()
    for batch in train_data:
        cur_contexts = _ctx_list
        if batch.data[0].shape[0] < len(_ctx_list):
            cur_contexts = cur_contexts[:batch.data[0].shape[0]]
        data = gluon.utils.split_and_load(batch.data[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        label = gluon.utils.split_and_load(batch.label[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        Ls = []
        with autograd.record():  # Start recording the derivatives
            for x_cur, y_cur in zip(data, label):
                L = loss(net(x_cur), y_cur)
                # store the loss and do backward after we have done forward
                # on all GPUs for better speed on multiple GPUs.
                Ls.append(L)
            for L in Ls:
                L.backward()
        trainer.step(batch.data[0].shape[0])
        
    scores_test = predict_scores(net, X_test, _batch_size, _ctx_list[0])
    predictions_test = np.argmax(scores_test, axis=1)
    accuracy = np.mean(predictions_test == y_test)
    print('[Epoch %d] accuracy=%f' % (e, accuracy))
```

Output:
```
[Epoch 0] accuracy=0.346900
[Epoch 0] accuracy=0.281900
```

**Comment by fedorzh** [comment]
I use Python 2 so the Issue https://github.com/apache/incubator-mxnet/issues/8658 shouldn't influence.

**Comment by sxjscience** [comment]
I think only one script is needed. We could run it multiple times to see if the results are the same.

**Comment by fedorzh** [comment]
To make sure this is not the initialization issue, I made `mx.init.One()`. With this, I wasn't able to reproduce on the whole dataset, but I was able to reproduce if trained on a subset:

```
n_batch=5000
random_selector = np.random.RandomState(0)
confidences = random_selector.rand(X.shape[0])
cert_idx = np.argsort(confidences)
selected_indices = cert_idx[:n_batch]
selected_indices = np.sort(selected_indices)
```

and 
```
train_data = mx.io.NDArrayIter(X[selected_indices, :], label=y[selected_indices], batch_size=_batch_size)
```

It is faster (only 5K examples to train on) and more clear (initialization should be deterministic) to work with this version

**Comment by roywei** [comment]
@sandeep-krishnamurthy please kindly add label: Bug, thanks!

**Comment by cjolivier01** [comment]
If I run the script by itself, without the second training run, I get the same result every time:
[Epoch 0] accuracy=0.100000

**Comment by cjolivier01** [comment]
Modified script per your second code comment (smaller test set):
```python
import mxnet as mx
from mxnet import nd, gluon, autograd, ndarray
import numpy as np
import random
from mxnet import profiler

# profiler.set_config(profile_symbolic=True, aggregate_stats=True, continuous_dump=True)
# profiler.set_state('run')

def transform(data, label):
    return [dat.astype(np.float32) for dat in data], [lab.astype(np.float32) for lab in label]


train_cifar_gluon = gluon.data.vision.CIFAR10(train=True, transform=transform)
test_cifar_gluon = gluon.data.vision.CIFAR10(train=False, transform=transform)

def convert_gluon_dataset_to_numpy(data):
    ds = data[:][0][0].shape
    X = np.empty((len(data[:][0]), ds[2], ds[0], ds[1]), dtype=np.float32)
    for i, example in enumerate(data[:][0]):
        X[i, :] = np.rollaxis(example.asnumpy(),2)
    y = np.array(data[:][1])
    return X, y

X, y = convert_gluon_dataset_to_numpy(train_cifar_gluon)
X_test, y_test = convert_gluon_dataset_to_numpy(test_cifar_gluon)


# In[2]:


def predict_scores(net, X_, batch_size, context):
    scores = None
    test_loaded = gluon.data.DataLoader(mx.nd.array(X_), batch_size, shuffle=False)
    for data in test_loaded:
        data = data.as_in_context(context)
        output = net(data).asnumpy()
        if scores is None:
            scores = output
        else:
            scores = np.append(scores, output, axis=0)
    return scores


# In[3]:


gpu_count = 1
_ctx_list = [mx.gpu(i) for i in range(gpu_count)]
_batch_size=64
epochs=1
_seed=42
_optimizer='sgd'
_learning_rate=0.1
_xavier_magnitude=2.
_momentum=0.9
_wd=0.0001
_nclasses=10


n_batch=5000
#n_batch=_batch_size * 2
random_selector = np.random.RandomState(0)
confidences = random_selector.rand(X.shape[0])
cert_idx = np.argsort(confidences)
selected_indices = cert_idx[:n_batch]
selected_indices = np.sort(selected_indices)

# ### Try 1

# In[4]:


random.seed(_seed)
mx.random.seed(_seed)
np.random.seed(_seed)


# In[5]:


net = gluon.model_zoo.vision.get_model('resnet34_v2', pretrained=False, classes=_nclasses, ctx=_ctx_list)

loss = gluon.loss.SoftmaxCrossEntropyLoss()


# In[6]:


net.collect_params().initialize(mx.init.Xavier(magnitude=_xavier_magnitude), ctx=_ctx_list, force_reinit=True)

trainer = gluon.Trainer(net.collect_params(), _optimizer,
                        optimizer_params=dict(learning_rate=_learning_rate, momentum=_momentum,
                                              wd=_wd),
                        kvstore='device' if len(_ctx_list) > 0 else 'local')

#train_data = mx.io.NDArrayIter(X, label=y, batch_size=_batch_size)
train_data = mx.io.NDArrayIter(X[selected_indices, :], label=y[selected_indices], batch_size=_batch_size)

for e in range(epochs):
    train_data.reset()
    for batch in train_data:
        cur_contexts = _ctx_list
        if batch.data[0].shape[0] < len(_ctx_list):
            cur_contexts = cur_contexts[:batch.data[0].shape[0]]
        data = gluon.utils.split_and_load(batch.data[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        label = gluon.utils.split_and_load(batch.label[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        Ls = []
        with autograd.record():  # Start recording the derivatives
            for x_cur, y_cur in zip(data, label):
                L = loss(net(x_cur), y_cur)
                # store the loss and do backward after we have done forward
                # on all GPUs for better speed on multiple GPUs.
                Ls.append(L)
            for L in Ls:
                L.backward()
        trainer.step(batch.data[0].shape[0])

    scores_test = predict_scores(net, X_test, _batch_size, _ctx_list[0])
    predictions_test = np.argmax(scores_test, axis=1)
    accuracy = np.mean(predictions_test == y_test)
    print('[Epoch %d] accuracy=%f' % (e, accuracy))


# ### Try 2

# In[7]:
# profiler.set_state('stop')
# print(profiler.dumps(True))

random.seed(_seed)
mx.random.seed(_seed)
np.random.seed(_seed)


# In[8]:


net = gluon.model_zoo.vision.get_model('resnet34_v2', pretrained=False, classes=_nclasses, ctx=_ctx_list)

loss = gluon.loss.SoftmaxCrossEntropyLoss()


# In[9]:


net.collect_params().initialize(mx.init.Xavier(magnitude=_xavier_magnitude), ctx=_ctx_list, force_reinit=True)

trainer = gluon.Trainer(net.collect_params(), _optimizer,
                        optimizer_params=dict(learning_rate=_learning_rate, momentum=_momentum,
                                              wd=_wd),
                        kvstore='device' if len(_ctx_list) > 0 else 'local')

#train_data = mx.io.NDArrayIter(X, label=y, batch_size=_batch_size)
train_data = mx.io.NDArrayIter(X[selected_indices, :], label=y[selected_indices], batch_size=_batch_size)

for e in range(epochs):
    train_data.reset()
    for batch in train_data:
        cur_contexts = _ctx_list
        if batch.data[0].shape[0] < len(_ctx_list):
            cur_contexts = cur_contexts[:batch.data[0].shape[0]]
        data = gluon.utils.split_and_load(batch.data[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        label = gluon.utils.split_and_load(batch.label[0], ctx_list=cur_contexts, batch_axis=0, even_split=False)
        Ls = []
        with autograd.record():  # Start recording the derivatives
            for x_cur, y_cur in zip(data, label):
                L = loss(net(x_cur), y_cur)
                # store the loss and do backward after we have done forward
                # on all GPUs for better speed on multiple GPUs.
                Ls.append(L)
            for L in Ls:
                L.backward()
        trainer.step(batch.data[0].shape[0])

    scores_test = predict_scores(net, X_test, _batch_size, _ctx_list[0])
    predictions_test = np.argmax(scores_test, axis=1)
    accuracy = np.mean(predictions_test == y_test)
    print('[Epoch %d] accuracy=%f' % (e, accuracy))
```

**Comment by cjolivier01** [comment]
I get the following (one GPU):
[Epoch 0] accuracy=0.100000
[Epoch 0] accuracy=0.100000

**Comment by cjolivier01** [comment]
Can you please supply build parameters that you use? ie CUDA, yes? CUDNN? MKL? etc.

**Comment by fedorzh** [comment]
Amazon p2.xlarge instance, cuda 9, cudnn7, no mkl

**Comment by cjolivier01** [comment]
Also, is this doen with the latest build?

**Comment by cjolivier01** [comment]
I didn't have CUDNN enabled. Trying that now...

**Comment by cjolivier01** [comment]
Ok, I can reproduce with CUDNN enabled...

**Comment by cjolivier01** [comment]
Is this supposed to take a really long time to run?  It takes many minutes...

**Comment by fedorzh** [comment]
@cjolivier01 
The function `convert_gluon_dataset_to_numpy` which converts gluon dataset to numpy takes awhile.

**Comment by sandeep-krishnamurthy** [comment]
I see that weights initialized are different for each run. To reproduce running below initialization and viewing the weights yields different weights. 

Taking a look.

```
net.collect_params().initialize(mx.init.Xavier(magnitude=_xavier_magnitude), ctx=_ctx_list, force_reinit=True)

net.collect_params()['resnetv21_conv0_weight'].data()
```

**Comment by sandeep-krishnamurthy** [comment]
Please ignore my previous comment. Thanks to Sina, for letting me know that seeding should happen just before the weight initialization to have deterministic behavior.
I verified that the initial state of all parameters across both the runs are same.
Sample code:

## Run 1
```
mx.random.seed(42)
w = gluon.parameter.Parameter('weight', shape = (3,3))
w.initialize(init=mx.initializer.Xavier(magnitude=2.0), ctx=mx.gpu(0))
```
```
[[ 0.53740764  0.4047128   0.09500974]
 [ 0.10734546  0.21412051  0.81077886]
 [ 0.39387202  0.0689261  -0.65652549]]
<NDArray 3x3 @cpu(0)>
```

## Run 2
```
mx.random.seed(42)
w.initialize(init=mx.initializer.Xavier(magnitude=2.0), ctx=mx.gpu(0), force_reinit=True)
w.data()
```

```
[[ 0.53740764  0.4047128   0.09500974]
 [ 0.10734546  0.21412051  0.81077886]
 [ 0.39387202  0.0689261  -0.65652549]]
<NDArray 3x3 @gpu(0)>
```

**Comment by sandeep-krishnamurthy** [comment]
Verified that the non-determinism is brought in by cuDNN backward algorithms for Conv and Pooling operators. Results are consistent on CPU and GPU without cuDNN.
This is expected behavior as per current MXNet implementation.
See here for an open issue where there is discussion on providing an option for determinism - https://github.com/apache/incubator-mxnet/issues/11341

Please note that, enabling deterministic cuDNN algorithm significantly slows the computation - https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility

Resolving the issue here in favor of new feature request tracked at https://github.com/apache/incubator-mxnet/issues/11341

Please reopen if closed in error.
